Tools/Scripts/hooks/pre-push

```diff
@@ -0,0 +1,356 @@
+            print_error("'{}' comes from an uncategorized remote, it's not safe to push it publically".format(commit))
+            return 1
+        # Security level of source exceeds security level of target
+        if categorization_class == 1 and security_level > (target_security_level or 0):
+            print_error("'{}' comes from a more secure remote than '{}'".format(commit, name))
+            if MODE == PUBLISH_MODE:
+                prompt_for.append(commit)
+            else:
+                return 1
+
+        # Unknown security level from original of cherry-pick
+        if categorization_class == 2 and security_level is None and not target_security_level:
+            print_error("'{}' cherry-picks a change from an uncategorized remote, it's not safe to push it publically".format(commit))
+            return 1
+        # Security level of original of a cherry-pick exceeds security level of target
+        if categorization_class == 2 and security_level > (target_security_level or 0):
+            print_error("'{}' cherry-picks a change from a more secure remote than '{}'".format(commit, name))
+            if MODE == DEFAULT_MODE:
+                prompt_for.append(commit)
+            else:
+                return 1
+
+        # Commit is redacted, but we're pushing it publically
+        if categorization_class == 3 and (MAX_LEVEL if security_level is None else security_level) > (target_security_level or 0):
+            print_error("'{}' fixes a redacted issue, but '{}' is a public remote".format(commit, name))
+            if MODE == DEFAULT_MODE:
+                prompt_for.append(commit)
+            else:
+                return 1
+
+        if VERBOSITY >= VERBOSE:
+            print('    {} is security level {} by class {}'.format(commit, security_level, categorization_class))
+
+    if prompt_for and not TTY:
+        print('No TTY available for prompting user, rejecting push')
+        return 1
+
+    if prompt_for:
+        branches = sorted(list(set([branch_for(commit) for commit in prompt_for])))
+        expected_response = 'publicize {}'.format(
+            branches[0] if len(branches) == 1
+            else '{} and {}'.format(', '.join(branches[:-1]), branches[-1]),
+        )
+        print('')
+        print('We have detected you are about to publish security content to a public remote')
+        print('Unless you are engaged in merge-back or fixing a security bug which has not shipped, you are likely making a mistake')
+        print('')
+        sys.stdout.write('Type "{}" to continue with publication: '.format(expected_response))
+        sys.stdout.flush()
+        response = TTY.readline().rstrip()
+        if response != expected_response:
+            print_error('User failed to explicitly approve potentially dangerous push')
+            return 1
+        print("User has explicitly approved potentially dangerous push, continuing...")
```

Comment:
A few small ideas here:

- Can we write to `TTY`, in case stdout is redirected somewhere weird?
- To help users detect typos, we could confirm when a response is recognized but does not match.
```suggestion
        TTY.write('Type "{}" to continue with publication: '.format(expected_response))
        TTY.flush()
        response = TTY.readline().rstrip()
        if response != expected_response:
            if response:
                print("Response does not match prompt")
            print_error('User failed to explicitly approve potentially dangerous push')
            return 1
        print("User has explicitly approved potentially dangerous push, continuing...")
```